std/algorithm/iteration.d

```diff
@@ -4280,3 +4286,220 @@ unittest
          [1, 2, 0],
          [2, 1, 0]]));
 }
+
+/**
+Finds the mean (colloquially known as the average) of a range. If $(D r) does
+not provide the $(D length) member, then this function will do element by
+element summation, rather than the more accurate methods provided by $(D sum).
+This function will return $(D real.nan) if the range is empty. This function is
+$(BIGOH r.length).
+
+Params:
+    r = an input range
+
+Returns:
+    the mean of r
+*/
+real mean(R)(R r)
+    if (isInputRange!R &&
+        !isInfinite!R &&
+        is(ElementType!R : real))
+{
+    if (r.empty)
+    {
+        return real.nan;
+    }
+
+    static if (hasLength!R)
+    {
+        if (r.length == 1)
+        {
+            return r.front;
+        }
+
+        return sum(r) / real(r.length);
+    }
+    else
+    {
+        import std.typecons : tuple;
+
+        auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
+            (tuple(size_t(0), 0.0), r);
+
+        return pair[1] / pair[0];
+    }
+}
+
+///
+@safe @nogc pure nothrow unittest
+{
+    import std.math : approxEqual, isNaN;
+
+    static immutable arr1 = [1, 2, 3];
+    static immutable arr2 = [1.5, 2.5, 12.5];
+
+    assert(mean(arr1) == 2);
+    assert(approxEqual(mean(arr2), 5.5));
+
+    assert(arr1[0 .. 0].mean.isNaN);
+}
+
+@safe pure nothrow unittest
+{
+    import std.internal.test.dummyrange : ReferenceInputRange;
+    import std.math : approxEqual;
+
+    auto r1 = new ReferenceInputRange!int([1, 2, 3]);
+    assert(r1.mean == 2);
+
+    auto r2 = new ReferenceInputRange!double([1.5, 2.5, 12.5]);
+    assert(approxEqual(r2.mean, 5.5));
+}
+
+/**
+Function to compute the quantile of a $(D SortedRange) given $(D q), which
+is a number between 0 and 1. The range MUST be a $(D SortedRange) that is sorted
+in ascending order. Otherwise this function will give incorrect results. If the
+range is empty, this function will return $(D real.nan). Is
+$(BIGOH r.length * q) for forward ranges and $(BIGOH 1) for random access ranges.
+
+Params:
+    r = a forward range with length or a random access range
+    q = where to calculate the quantile, must be $(D q > 0 && q < 1)
+
+Returns:
+    the quantile of r given q
+*/
+real quantile(R)(R r, real q)
+    if (isInstanceOf!(SortedRange, R) &&
+        is(ElementType!R : real) &&
+        isForwardRange!R &&
+        hasLength!R &&
+        !isInfinite!R)
+in
+{
+    assert(q > 0 && q < 1);
```

Comment:
> because the compiler can remove them, without you knowing.

And that's fine, because they'll be in release mode which must be explicitly activated and they should have run it with contracts on at least once while developing. This seems to be perfectly acceptable for a lot of the functions in other modules, especially `std.range` which makes heavy use of contracts.